pkg/scheduler/eventhandlers/reservation_handler.go

```diff
@@ -0,0 +1,132 @@
+/*
+Copyright 2022 The Koordinator Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package eventhandlers
+
+import (
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/klog/v2"
+	"k8s.io/kubernetes/pkg/scheduler"
+	"k8s.io/kubernetes/pkg/scheduler/framework"
+
+	schedulingv1alpha1 "github.com/koordinator-sh/koordinator/apis/scheduling/v1alpha1"
+	"github.com/koordinator-sh/koordinator/pkg/scheduler/frameworkext"
+	"github.com/koordinator-sh/koordinator/pkg/scheduler/plugins/reservation"
+)
+
+var assignedPodDelete = framework.ClusterEvent{Resource: framework.Pod, ActionType: framework.Delete, Label: "AssignedPodDelete"}
+
+func AddScheduleEventHandler(sched *scheduler.Scheduler, extendedHandle frameworkext.ExtendedHandle) {
+	reservationInformer := extendedHandle.KoordinatorSharedInformerFactory().Scheduling().V1alpha1().Reservations().Informer()
+	reservationInformer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			r, ok := obj.(*schedulingv1alpha1.Reservation)
+			if ok {
+				addReservation(sched, r)
+			} else {
+				klog.Errorf("reservation informer add func parse obj failed, obj %T", obj)
+			}
+		},
+		UpdateFunc: func(oldObj, newObj interface{}) {
+			oldR, oldOK := oldObj.(*schedulingv1alpha1.Reservation)
+			newR, newOK := newObj.(*schedulingv1alpha1.Reservation)
+			if !oldOK || !newOK {
+				klog.Errorf("unable to convert object to *schedulingv1alpha1.Reservation, old %T, new %T", oldObj, newObj)
+				return
+			}
+			updateReservation(sched, oldR, newR)
+		},
+		DeleteFunc: func(obj interface{}) {
+			switch t := obj.(type) {
+			case *schedulingv1alpha1.Reservation:
+				deleteReservation(sched, t)
+				return
+			case cache.DeletedFinalStateUnknown:
+				r, ok := t.Obj.(*schedulingv1alpha1.Reservation)
+				if ok {
+					deleteReservation(sched, r)
+					return
+				}
+			}
+			klog.Errorf("reservation informer delete func parse obj failed, obj %T", obj)
+		},
+	})
+}
+
+func addReservation(sched *scheduler.Scheduler, r *schedulingv1alpha1.Reservation) {
+	if r == nil || r.Spec.Template == nil {
+		klog.Errorf("failed to add reservation for invalid reservation %v", r)
+		return
+	}
+	reservePod := reservation.GetReservePod(r)
+	// TBD: should the reservation with nodeName specified enqueue?
+	if reservePod.Status.Phase != schedulingv1alpha1.ReasonReservationExpired {
```

Comment:
check the reservePod.Status.Phase whether is ReasonReservationExpired? is it a typo?
I guess you want to check reservation.status.phase.

Reply:
nice catch